### ISSUE TITLE
GUI Effect checker: annotate lambdas

### ISSUE THREAD

**Issue by vlsi** [typetools/checker-framework#5821]
It is not clear how UI effects could be annotated for lambdas.

Is there a way I could annotate lambdas so they are treated as if they were `@UIEffect`?

For instance:

```java
val usage() {
    // error: [call.ui] Calling a method with UIEffect effect from a context limited to SafeEffect effects.
    runEffect(component -> component.setForeground(Color.BLACK));
}


@UIEffect
void runEffect(Callable<JComponent> effect) {
    // ..
}
```

See sample in JMeter:

Call site:
* https://github.com/apache/jmeter/blob/0a65d7ae147aa5464bfb8dc997604f81a8d0749e/src/jorphan/src/main/java/org/apache/jorphan/gui/JFactory.java#L145-L147

Declaration:
* https://github.com/apache/jmeter/blob/0a65d7ae147aa5464bfb8dc997604f81a8d0749e/src/jorphan/src/main/java/org/apache/jorphan/gui/DynamicStyle.java#L63C74-L69

**Comment by vlsi** [comment]
```java
void runEffect(@UI Callable<JComponent> effect) {
```

produces

```
jmeter/src/jorphan/src/main/java/org/apache/jorphan/gui/DynamicStyle.java:64: error: [annotations.on.use] invalid type: annotations [@UI] conflict with declaration of type java.util.function.Consumer
    public <T extends JComponent> T withDynamic(T component, @UI Consumer<T> onUpdateUi) {
```

**Comment by smillst** [comment]
I think you have the right idea, but `Callable` needs to be annotated as `@PolyUI` and its method `@PolyUIEffect`.  I think the annotated JDK is missing a lot of these annotations for functional interfaces.  Here's an example that works:

```
import javax.swing.JComponent;
import javax.swing.JLabel;
import javax.swing.UIManager;
import org.checkerframework.checker.guieffect.qual.PolyUI;
import org.checkerframework.checker.guieffect.qual.PolyUIEffect;
import org.checkerframework.checker.guieffect.qual.UI;

public class GuiEffectLambdas {

  @PolyUI
  interface Consumer<Q> {
    @PolyUIEffect
    void accept(Q q);
  }

  public JLabel error(JLabel component) {
    return withDynamic(component, c -> {
      c.setFont(UIManager.getFont("JMeterUIDefaults.LABEL_ERROR_FONT"));
      c.setForeground(UIManager.getColor("JMeterUIDefaults.LABEL_ERROR_FOREGROUND"));
    });
  }


  public <T extends JComponent> T withDynamic(T component, @UI Consumer<T> onUpdateUi) {
    throw new RuntimeException();
  }
}
```

**Comment by vlsi** [comment]
Thanks, it helped, except I replaced `@PolyUI` with `@PolyUIType`.
I've annotated one of JMeter's modules.

It looks like `PropertyChangeListener` concept is not annotated in a coherent style.
I think the way to go should be:
* `PropertyChangeListener` is `@PolyUIType`
* Methods like `Component#addPropertyChangeListener(String name, PropertyChangeListener listener)` should be `@SafeEffect Component#addPropertyChangeListener(String name, @UI PropertyChangeListener listener)`

Currently `addPropertyChangeListener` rejects `@UI PropertyChangeListeners`, and it assumes `firePropertyChange` has `@UIEffect` which is not aligned.

However, let me annotate more code so I gather a better set of stubs.
If you like to peek at the current set of stubs, you might check https://github.com/apache/jmeter/pull/5871
